Blog post: Strimzi based Apache Kafka infrastructure as a Java code (#311)
strimzi/strimzi.github.io: _posts/2022-08-02-stimzi-based-infrastructure-as-a-java-code.md

---
layout: post

> title: "Strimzi based Apache Kafka infrastructure as a Java code"

Reviewer: ```suggestion
title: "Strimzi-based Apache Kafka infrastructure as Java code"
```

date: 2022-08-01
author: jakub_scholz
---

When you say _Infrastructure as Code (IaC)_, people usually imagine tools such as Terraform, Ansible or Amazon AWS CloudFormation.
All these tools use some kind of configuration or markup language to write the code.

> Terraform uses the HCL language, Ansible uses YAML and AWS CLoudFormation can use either YAML or JSON.

Reviewer 2: nit: `CLoudFormation` -> `CloudFormation`

Strimzi supports this pattern as well.

> You describe your Strimzi based Apache Kafka infrastructure in the YAML format when you create the Kubernetes custom resources.

Reviewer: ```suggestion
You describe your Strimzi-based Apache Kafka infrastructure in the YAML format when you create the Kubernetes custom resources.
```

And Kubernetes together with the Strimzi operators take care of transforming this _code_ into the actual Apache Kafka brokers, users, topics or for example connectors.

> But what if you want to use a full-fledged programming language for your infrastructure ... for example Java?

Reviewer: Should we motivate more (here in the lede) why you might want to do it with Java? Or at least say we'll answer the "Why" question in the main content?

Author: I went for the second option you suggested and reworded it to make it more clear the _why_ is covered in the blog post as well. 

TBH, I think this is a bit weird. We typically use the lede in the blog posts. But we do not use it anywhere in our design. The blog post listing shows titles only, the RSS syndication pushes full articles. So I'm actually not sure why do we have the `<!--more-->` marker there. But I kept it here since we follow it in most of our blog posts. Maybe (long-term) this is something to think about and either use it better in the design or stop using it in the blog post.

Another example might be when you want to manage your Kafka infrastructure directly from some integration or system tests written in Java.
You can create a Kafka cluster or prepare Kafka topics directly in the setup phase before each of your tests.

### Strimzi `api` module

Strimzi is written in Java.
The main interface used by Strimzi users are the Kubernetes Custom Resource Definitions (CRDs) and Custom Resources (CRs).
They constitute the main Strimzi API.
But we of course do not write the CRDs by hand.
Instead, we have Java classes representing the custom resources and their definition.
These Java classes live in the `api` module.

They are used in multiple different ways.
The YAML files with the CRD definitions are automatically generated based on them.
The [API reference part of our documentation](https://strimzi.io/docs/operators/latest/full/configuring.html#api_reference-str) is automatically generated from these files.
And they are (together with some automatically generated builder and fluent classes) also used by the Strimzi operators to work with the custom resources. 

> When the operator queries the Kubernetes API or receives some event from it, it will be formatted in a JSON format.

Reviewer: ```suggestion
When the operator queries the Kubernetes API or receives some event from it, it will be serialize in JSON format.
```

The operators will decode the custom resources from JSON into regular Java objects.
And when the operator needs to modify any of the custom resources, it will again use the Java objects to do so.
And they will be converted from the Java format into JSON again before updating them in the Kubernetes API.
This all happens automatically with the help of few libraries:

> * The [Sundrio code generation toolkit](https://github.com/sundrio/sundrio) toolkit is used to generate the builder and fluent interfaces for the Strimzi custom resources

Reviewer 3: ```suggestion
* The [Sundrio code generation toolkit](https://github.com/sundrio/sundrio) is used to generate the builder and fluent interfaces for the Strimzi custom resources
```

* The [Fabric8 Kubernetes client](https://github.com/fabric8io/kubernetes-client) is used to communicate with the Kubernetes API and together with the [FasterXML Jackson library](https://github.com/FasterXML/jackson) does the decoding and encoding between JSON or YAML and Java.

And the best thing about this is that its use is not limited only to the Strimzi operators.
If you want, you can use it as well!

### Using the `api` module

With every Strimzi release, we publish the `api` library to the [Maven Central repository](https://mvnrepository.com/artifact/io.strimzi/api).
So you can easily include it in your Java project.

> For example, if you use Maven to build you project, you can just add the following dependency:

Reviewer 3: ```suggestion
For example, if you use Maven to build your project, you can just add the following dependency:
```

This code already updates the custom resource inside the Kubernetes API.
The new YAML corresponding to the Java code would look like this:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: <NAME>
  namespace: <NAMESPACE>
spec:
  kafka:
    replicas: 3
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
    storage:
      type: ephemeral
    config:
      auto.create.topics.enable: "false"
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    topicOperator: {}
    userOperator: {}
```

Once the resource is updated in Kubernetes, Strimzi will need to roll the Kafka cluster to apply the change.
We can once again wait for it to complete:

```java
Crds.kafkaOperation(client).inNamespace(NAMESPACE).withName(NAME).waitUntilCondition(k -> {
    if (k.getStatus() != null && k.getStatus().getConditions() != null) {
        return updatedKafka.getMetadata().getGeneration() == k.getStatus().getObservedGeneration()
                && k.getStatus().getConditions().stream().anyMatch(c -> "Ready".equals(c.getType()) && "True".equals(c.getStatus()));
    } else {
        return false;

>     }

Reviewer: We should encapsulate this and provide it from the `api` module. e.g. `waitUntilCondition(Conditions.reconciled(), 5, TimeUnit.MINUTES)`

Reviewer: Maybe we could open a "good start" issue for this, and link to it from the blog post, to see if someone wants to contribute it?

Author: Good idea. I opened https://github.com/strimzi/strimzi-kafka-operator/issues/7138 for this.

### Other custom resources

In the previous sections, we looked at how to create, update and delete a Strimzi-based Apache Kafka cluster from a Java application.
The same can be of course used for all other Strimzi custom resources.
So you are not limited only to Kafka clusters.

### Complete examples

The full source codes used in this blog post can be found in my [GitHub repository](https://github.com/scholzj/strimzi-api-examples).
It does not contain examples for all custom resources, but if you want to add some more, feel free to open PR or an issue.
The repository also contains an example of how you can install the Strimzi Cluster Operator directly from a Java application.
Feel free to reuse and share these examples as needed.

### What about other languages?

In this blog post, we focused only on Java.
What if you are using some other programming language?

For example, for Golang, there are several projects which allow you to generate Go structures from the YAML files to help you with managing your Kubernetes resources.

> One of them is for example the [Reverse Kube Resource](https://github.com/scholzj/strimzi-api-examples).

Reviewer 4: wrong link here?
